nixos/modules/services/mail/dovecot.nix

```diff
@@ -1,19 +1,90 @@
 { config, lib, pkgs, ... }:
 
 let
-  inherit (lib) attrValues concatMapStringsSep concatStrings
-    concatStringsSep flatten imap1 literalExpression mapAttrsToList
-    mkEnableOption mkIf mkOption mkRemovedOptionModule optional optionalAttrs
-    optionalString singleton types mkRenamedOptionModule nameValuePair
-    mapAttrs' listToAttrs filter;
-  inherit (lib.strings) match;
+  inherit (lib) attrValues concatMapStringsSep concatStrings concatStringsSep
+    flatten imap1 isBool isDerivation isInt isList isPath isString
+    literalExpression mapAttrsToList mkEnableOption mkIf mkOption
+    mkRemovedOptionModule optional optionalAttrs optionalString singleton
+    types mkRenamedOptionModule nameValuePair mapAttrs' listToAttrs filter;
+  inherit (lib.generators) toPretty;
+  inherit (lib.lists) findFirstIndex length sublist reverseList;
+  inherit (lib.strings) escape match splitString;
 
   cfg = config.services.dovecot2;
   dovecotPkg = pkgs.dovecot;
 
   baseDir = "/run/dovecot2";
   stateDir = "/var/lib/dovecot";
 
+  /* Helper: Format string in way that may strech over multiple
+   * lines in a Dovecot configuration file
+   *
+   * This is a little complicated since Dovecot does not like when the last
+   * line is empty, so we first have to truncate all trailing empty lines
+   * before when can join them using the continuation line syntax.
+   */
+  formatDovecotMultiline =
+    string:
+      let
+        # Split input into line list
+        lines = splitString "\n" string;
+        # Find index of last non-empty line in line list
+        lastContentIndex = length lines - (findFirstIndex (
+          line: match "[[:space:]]*" line == null  # <=> Line is not empty
+        ) (length lines) (reverseList lines));
+        # Drop empty trailing lines from line list
+        trimmedLines = sublist 0 lastContentIndex lines;
+      in
+        # Reassemble content without trailing lines but with continuation markers
+        concatStringsSep " \\\n" trimmedLines;
+
+  /* Helper: Apply string escaping and quotation to Dovecot string if needed
+   *
+   * Dovecot is sometimes touchy about strings being quoted, so this only
+   * quotes the string when it is actually needed.
+   */
+  escapeDovecotString = string:
+    let
+      escapedString = escape ["\\" "\""] string;
+    in   if match "[[:space:]]*<.*|.*[[:space:]#\"\\].*|" string != null
+    then "\"${formatDovecotMultiline escapedString}\""
+    else string;
+
+  /* Helper: Convert a Dovecot value to a string
+   *
+   * Strings are escaped except for config variable interpolation, as such
+   * this function is unsuitable for performing in-Dovecot variable
+   * substitutions, however these can also easily be done using Nix instead.
+   *
+   * Reference: https://doc.dovecot.org/settings/types/
+   */
+  mkDovecotValue =
+    value: (
+      if isString value || isPath value || isDerivation value
+      then
+        escapeDovecotString (toString value)
```

Comment:
With this, you'll have to pass the list as-is to pluginSettings for sieve.plugins, sieve.extensions and sieve.globalExtensions, otherwise it will quote the values:

```
plugin {
  [...]
  sieve_extensions = "+notify +imapflags +vnd.dovecot.filter"
  sieve_global_extensions = "+vnd.dovecot.environment +vnd.dovecot.pipe"
  sieve_pipe_bin_dir = /nix/store/9g0l06qac224hd2m0ixj0mhf25mh0662-sieve-pipe-bins
  sieve_plugins = "sieve_imapsieve sieve_extprograms"
} 
```

Reply:
Is this about having to write

```nix
services.dovecot = {
    pluginSettings = {
        sieve_extensions = ["+notify" "+imapflags" "+vnd.dovecot.filter"];
    };
};
```

vs

```nix
services.dovecot = {
    pluginSettings = {
        sieve_extensions = "+notify +imapflags +vnd.dovecot.filter";
    };
};
```

?

If so, that is intended behaviour since the former denotes a list of strings (which this is), while the latter denotes a since string containing spaces. (Having it declared as a list also means others can later use `mkBefore`/`mkAfter` in their own configs to easily extend it without conflicts or `mkForce`.)

Dovecot isn’t very strict about these type differences in practice, so both of the above actually work. On the hand, a configuration value of `sieve_extensions = "+notify" "+imapflags" "+vnd.dovecot.filter"` *doesn’t* work, even though the documentation says it should. As far as I can tell, all of those “should work but doesn’t” cases boil down to being space-separated words which the regex mentioned in the previous comment gracefully handles by making sure the quoting is only applied when it is actually needed due to the contents containing “special characters” (spaces, quotation marks, backslashes, file redirects, comment signs, empty strings). I actually have an extended version of what I’m proposing here (uses the same code also for global assignments, mailboxes, services and service listeners) in use and as long as one follows the only-quote-when-needed rule, even complex configurations are successfully generated and parsed by Dovecot with this.

Reply:
No, this is about [the settings that I mentioned](https://github.com/NixOS/nixpkgs/blob/522d660f2512166ee7320875d9dde072940257b0/nixos/modules/services/mail/dovecot.nix#L551-L553) setting your second variant instead of the first. Even if it works (tbh I didn't assume it would) I think it would be good to change it.

Reply:
I guess with this at least sieve.plugins could be removed since it just passes the value through directly.
If you do that though then I'd make pluginSettings a submodule with freeformType like [here](https://github.com/NixOS/nixpkgs/blob/nixos-unstable/nixos/modules/config/nix.nix#L173) so the description and example can be kept.